crates/relayer/relayer-components/src/relay/impls/packet_clearers/packets.rs

```diff
@@ -0,0 +1,75 @@
+use cgp_core::async_trait;
+
+use crate::chain::traits::queries::packet_acknowledgements::CanQueryPacketAcknowledgements;
+use crate::chain::traits::queries::packet_commitments::CanQueryPacketCommitments;
+use crate::chain::traits::queries::send_packets::CanQuerySendPackets;
+use crate::chain::traits::queries::unreceived_acks_sequences::CanQueryUnreceivedAcksSequences;
+use crate::chain::traits::queries::unreceived_packet_sequences::CanQueryUnreceivedPacketSequences;
+use crate::chain::types::aliases::{ChannelIdOf, PortIdOf};
+use crate::relay::traits::chains::{CanRaiseRelayChainErrors, HasRelayChains};
+use crate::relay::traits::packet_clearer::PacketClearer;
+use crate::relay::traits::packet_relayer::CanRelayPacket;
+use crate::runtime::traits::runtime::HasRuntime;
+use crate::runtime::traits::task::{CanRunConcurrentTasks, Task};
+
+use super::ack::ClearAckPackets;
+use super::receive_packet::ClearReceivePackets;
+
+pub struct ClearAllPackets;
+
+pub struct RelayPacketTask<Relay>
+where
+    Relay: HasRelayChains,
+{
+    pub relay: Relay,
+    pub packet: Relay::Packet,
+}
+
+#[async_trait]
+impl<Relay> Task for RelayPacketTask<Relay>
+where
+    Relay: CanRelayPacket,
+{
+    async fn run(self) {
+        let _ = self.relay.relay_packet(&self.packet).await;
+    }
+}
+
+#[async_trait]
+impl<Relay> PacketClearer<Relay> for ClearAllPackets
+where
+    Relay: Clone + CanRelayPacket + HasRuntime + CanRaiseRelayChainErrors,
+    Relay::DstChain: CanQueryUnreceivedPacketSequences<Relay::SrcChain>
+        + CanQueryUnreceivedAcksSequences<Relay::SrcChain>
+        + CanQueryPacketAcknowledgements<Relay::SrcChain>,
+    Relay::SrcChain:
+        CanQueryPacketCommitments<Relay::DstChain> + CanQuerySendPackets<Relay::DstChain>,
+    Relay::Runtime: CanRunConcurrentTasks,
+{
+    async fn clear_packets(
+        relay: &Relay,
+        src_channel_id: &ChannelIdOf<Relay::SrcChain, Relay::DstChain>,
+        src_port_id: &PortIdOf<Relay::SrcChain, Relay::DstChain>,
+        dst_channel_id: &ChannelIdOf<Relay::DstChain, Relay::SrcChain>,
+        dst_port_id: &PortIdOf<Relay::DstChain, Relay::SrcChain>,
+    ) -> Result<(), Relay::Error> {
+        ClearReceivePackets::clear_packets(
```

Comment:
@soareschen Did I understand correctly earlier that you were suggesting that we run these two tasks concurrently?

Reply:
Yup. We can implement this much simpler as follows:

```rust
use alloc::vec;

use crate::chain::traits::types::ibc::HasIbcChainTypes;
use crate::chain::types::aliases::{ChannelIdOf, PortIdOf};
use crate::relay::traits::chains::HasRelayChains;
use crate::relay::traits::packet_clearer::PacketClearer;
use crate::runtime::traits::runtime::HasRuntime;
use crate::runtime::traits::task::{CanRunConcurrentTasks, Task};

use super::ack::ClearAckPackets;
use super::receive_packet::ClearReceivePackets;

pub struct ClearAllPackets;

pub enum ClearOption {
    Receive,
    Ack,
}

pub struct RunPacketClearer<Relay>
where
    Relay: HasRelayChains,
{
    pub relay: Relay,
    pub src_channel_id: ChannelIdOf<Relay::SrcChain, Relay::DstChain>,
    pub src_port_id: PortIdOf<Relay::SrcChain, Relay::DstChain>,
    pub dst_channel_id: ChannelIdOf<Relay::DstChain, Relay::SrcChain>,
    pub dst_port_id: PortIdOf<Relay::DstChain, Relay::SrcChain>,
    pub clear_option: ClearOption,
}

impl<Relay> Task for RunPacketClearer<Relay>
where
    Relay: HasRelayChains,
    ClearReceivePackets: PacketClearer<Relay>,
    ClearAckPackets: PacketClearer<Relay>,
{
    async fn run(self) {
        let _ = match self.clear_option {
            ClearOption::Receive => {
                ClearReceivePackets::clear_packets(
                    &self.relay,
                    &self.src_channel_id,
                    &self.src_port_id,
                    &self.dst_channel_id,
                    &self.dst_port_id,
                )
                .await
            }
            ClearOption::Ack => {
                ClearAckPackets::clear_packets(
                    &self.relay,
                    &self.src_channel_id,
                    &self.src_port_id,
                    &self.dst_channel_id,
                    &self.dst_port_id,
                )
                .await
            }
        };
    }
}

impl<Relay, SrcChain, DstChain> PacketClearer<Relay> for ClearAllPackets
where
    Relay: Clone + HasRuntime + HasRelayChains<SrcChain = SrcChain, DstChain = DstChain>,
    SrcChain: HasIbcChainTypes<DstChain>,
    DstChain: HasIbcChainTypes<SrcChain>,
    Relay::Runtime: CanRunConcurrentTasks,
    ClearReceivePackets: PacketClearer<Relay>,
    ClearAckPackets: PacketClearer<Relay>,
    SrcChain::ChannelId: Clone,
    SrcChain::PortId: Clone,
    DstChain::ChannelId: Clone,
    DstChain::PortId: Clone,
{
    async fn clear_packets(
        relay: &Relay,
        src_channel_id: &SrcChain::ChannelId,
        src_port_id: &SrcChain::PortId,
        dst_channel_id: &DstChain::ChannelId,
        dst_port_id: &DstChain::PortId,
    ) -> Result<(), Relay::Error> {
        let receive_packet_task = RunPacketClearer {
            relay: relay.clone(),
            src_channel_id: src_channel_id.clone(),
            src_port_id: src_port_id.clone(),
            dst_channel_id: dst_channel_id.clone(),
            dst_port_id: dst_port_id.clone(),
            clear_option: ClearOption::Receive,
        };

        let ack_packet_task = RunPacketClearer {
            relay: relay.clone(),
            src_channel_id: src_channel_id.clone(),
            src_port_id: src_port_id.clone(),
            dst_channel_id: dst_channel_id.clone(),
            dst_port_id: dst_port_id.clone(),
            clear_option: ClearOption::Ack,
        };

        relay
            .runtime()
            .run_concurrent_tasks(vec![receive_packet_task, ack_packet_task])
            .await;

        Ok(())
    }
}
```